### PR TITLE
Intersection API Unit Testing #3

### PR DESCRIPTION
**Pull request by jacob6838**
# PR Details

## Description

Writing additional unit tests for the Intersection API, bringing the total instruction and line coverage to 85%. 
image

This involved new tests related to the following areas:
- Controllers
- Accessors
- Properties
- Tasks

## How Has This Been Tested?

These tests can be run using:
```sh
cd services/intersection-api/api
mvn clean install
mvn verify
```

And the coverage can be viewed by opening the following file in a broser:
services/intersection-api/api/target/site/jacoco/index.html

## Types of changes

- [x] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] My changes require new environment variables:
  - [ ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [x] My changes require updates to the documentation:
  - [x] I have updated the documentation accordingly.
- [x] My changes require updates and/or additions to the unit tests:
  - [x] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

### CODE REVIEW

**Comment by drewjj** [comment]
Having issues running the tests. Seems like Mockito is not happy. I get this when I run `mvn clean install`:

image

**Comment by payneBrandon** [comment]
`mvn clean compile` still seems to work, though I do see a couple warnings about unchecked conversions that we'll want to look at: 
image

**Comment by jacob6838** [comment]
I will check on those warnings. The issue that Drew noted only occurs when running with Java 22 or above

**Comment by payneBrandon** [comment]
I don't think we have it in here now, but can we get a pipeline to run the verify command on PR? I'm seeing some errors with java 21 when running tests, but a pipeline would surface those sooner:
image

**Comment by jacob6838** [comment]
@payneBrandon that issue is from not having the docker deamon running. This is an unfortuate requirement for the mock database that we use in some of the tests. I want to find a better way to error out for that, but haven't dug into it yet.
We have a github action runing called build_intersection_api, which did pass

**Comment by payneBrandon** [comment]
> @payneBrandon that issue is from not having the docker deamon running. This is an unfortuate requirement for the mock database that we use in some of the tests. I want to find a better way to error out for that, but haven't dug into it yet. We have a github action runing called build_intersection_api, which did pass

Ah interesting, running with docker enabled does succeed. I don't see a section in the README about testing, let's add that into there with the docker daemon requirement to make it more explicit what's needed.

**Comment by jacob6838** [comment]
> > @payneBrandon that issue is from not having the docker deamon running. This is an unfortuate requirement for the mock database that we use in some of the tests. I want to find a better way to error out for that, but haven't dug into it yet. We have a github action runing called build_intersection_api, which did pass
> 
> Ah interesting, running with docker enabled does succeed. I don't see a section in the README about testing, let's add that into there with the docker daemon requirement to make it more explicit what's needed.

I have a section in the services/intersection-api/api/README.md, but there are a few other READMEs above that in the project. I'll add a section to services/intersection-api as well

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java [hunk 1]
```diff
@@ -55,133 +55,161 @@
...
+                Page<ConnectionOfTravelAssessment> page = repository.findLatest(intersectionID, startTime, endTime);
+
+                assertThat(page.getContent()).hasSize(1);
+                assertThat(page.getContent().get(0).getIntersectionID()).isEqualTo(intersectionID);
+                verify(mongoTemplate).findOne(any(Query.class), eq(ConnectionOfTravelAssessment.class),
+                                eq("CmConnectionOfTravelAssessment"));
+        }
+
+        @Test
+        void testAdd() {
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java]: question: do we need any edge case testing here? ie what happens if we try to insert a null object?

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/ConfigControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+    @Test
+    void testDefaultConfigException() {
+        DefaultConfig<String> inputConfig = new DefaultConfig<>();
+        inputConfig.setKey("testKey");
+        inputConfig.setValue("newValue");
+
+        String resourceURL = "http://localhost/config/default/testKey";
+        when(props.getCmServerURL()).thenReturn("http://localhost");
+        when(restTemplate.getForEntity(resourceURL, DefaultConfig.class))
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/ConfigControllerTest.java:131]: question: in our other tests we use the `eq()` method when checking the resourceURL and DefaultConfig.class values here. Is there a reason we're not doing the same here? Is there one method that's considered best practice?

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/IntersectionControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        String username = "testuser";
+        try (MockedStatic<PermissionService> mockedStatic = Mockito.mockStatic(PermissionService.class)) {
+            mockedStatic.when(() -> PermissionService.getUsername(any())).thenReturn(username);
+            when(postgresService.getAllowedIntersectionIdsByEmail(username))
+                    .thenReturn(Collections.singletonList(2));
+
+            ResponseEntity<List<IntersectionReferenceData>> response = controller.getIntersections(null, false);
+            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+            assertThat(response.getBody()).contains(intersection);
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/IntersectionControllerTest.java]: suggestion: we're configuring a single intersection to return for the given email, it would make sense to ensure our return body has a single object which _is_ the intersection, rather than contains the intersection (and possibly other data).

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Test
+    void testBroadcastSpatNullIntersectionId() {
+        ProcessedSpat spat = mock(ProcessedSpat.class);
+        when(spat.getIntersectionId()).thenReturn(null);
+
+        controller.broadcastSpat(spat);
+
+        // Should not send if intersectionID == -1
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java:73]: suggestion: checking against -1 would be a good separate unit test, but in this one we're testing null

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Test
+    void testBroadcastSpat() throws JsonProcessingException {
+        ProcessedSpat spat = mock(ProcessedSpat.class);
+        when(spat.getIntersectionId()).thenReturn(42);
+
+        controller.broadcastSpat(spat);
+
+        verify(brokerMessagingTemplate).convertAndSend(startsWith("/live/42/processed-spat"), anyString());
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java:52]: question: we're calling `broadcastSpat` but checking the topic has "processed-spat" in it which is misleading. is this method supposed to just broadcast the processed-spat messages, and should it be named accordingly?

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompSessionControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    void testHandleSessionConnectEvent_NullSessionIdThrows() {
+        MessageHeaders headers = new MessageHeaders(new HashMap<>());
+        Message<byte[]> message = mock(Message.class);
+        when(message.getHeaders()).thenReturn(headers);
+
+        AbstractSubProtocolEvent event = mock(SessionConnectEvent.class);
+        when(event.getMessage()).thenReturn(message);
+
+        try {
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompSessionControllerTest.java]: question: in many of our other tests we use the `assertThrows` rather than try/catch without an assert. Is there a reason for the swap?

**Review by payneBrandon — CHANGES_REQUESTED** [review]
A few questions so far

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/events/BsmMessageCountProgressionEventRepositoryImplTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import org.mockito.*;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Query;
+import us.dot.its.jpo.conflictmonitor.monitor.models.events.BsmMessageCountProgressionEvent;
+import us.dot.its.jpo.ode.api.accessors.events.bsm_message_count_progression_event.BsmMessageCountProgressionRepositoryImpl;
+
+import java.util.List;
```
Comment (payneBrandon) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/events/BsmMessageCountProgressionEventRepositoryImplTest.java]: suggestion: remove unused import

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java (hunk 1 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java]: Calling mongoTemplate.insert(null, collectionName) will throw a NullPointerException - I am ok with that being the failure mode. Because of the simplicity of the add method in that it just flatly calls mongoTemplate.insert, I am happy with the amount of testing for this method

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/ConfigControllerTest.java (hunk 2 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/ConfigControllerTest.java:131]: There is honestly not much of a difference, they do the same thing in this case. The matchers (eq, anyString, ...) are useful when you want a more generic match, and they can be avoided when exact values are used. I am not aware of any best practices, just case specific

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/IntersectionControllerTest.java (hunk 3 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/intersections/IntersectionControllerTest.java]: Updated to verify length and value

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java (hunk 4 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java:73]: Added test

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java (hunk 5 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompControllerTest.java:52]: Re-named to broadcastProcessedSpat and broadcastProcessedMap

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompSessionControllerTest.java (hunk 6 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/controllers/live/StompSessionControllerTest.java]: updated to use assertThrows

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/events/BsmMessageCountProgressionEventRepositoryImplTest.java (hunk 7 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/events/BsmMessageCountProgressionEventRepositoryImplTest.java]: removed

File: services/intersection-api/api/README.md [hunk 8]
```diff
@@ -77,6 +121,22 @@ To run unit tests, run the following command (ensure docker engine is running):
 mvn test
```
Comment (payneBrandon) [inline services/intersection-api/api/README.md:121]: blocker: Note that the maven commands currently don't work unless specifying the settings.xml to point to the Neaera repository. This is underway to pull into USDOT and CDOT repos.

Comment (jacob6838) [reply services/intersection-api/api/README.md:121]: This is ready for re-review! THe FFM repo is now on usdot https://github.com/usdot-jpo-ode/j2735-ffm-java. For the current point in the release process, you will want to add usdot-jpo-ode to your .m2 settings.xml in addition to the cdot-cv organization
```xml
<?xml version="1.0" encoding="UTF-8"?>
<settings>
    <activeProfiles>
        <activeProfile>default</activeProfile>
    </activeProfiles>
    <servers>
        <server>
            <id>github-usdot</id>
            <username>usdot-jpo-ode</username>
            <password>MAVEN_GITHUB_TOKEN</password>
        </server>
        <server>
            <id>github-cdot</id>
            <username>cvmanager_intersection_api</username>
            <password>MAVEN_GITHUB_TOKEN</password>
        </server>
    </servers>
    <profiles>
        <profile>
            <id>default</id>
            <repositories>
                <repository>
                    <id>github-usdot</id>
                    <name>GitHub Apache Maven Packages USDOT JPO ODE</name>
                    <url>https://maven.pkg.github.com/usdot-jpo-ode/*</url>
                    <snapshots>
                        <enabled>false</enabled>
                    </snapshots>
                </repository>
                <repository>
                    <id>github-cdot</id>
                    <name>GitHub Apache Maven Packages CDOT CV</name>
                    <url>https://maven.pkg.github.com/cdot-cv/*</url>
                    <snapshots>
                        <enabled>false</enabled>
                    </snapshots>
                </repository>
            </repositories>
        </profile>
    </profiles>
</settings>
```

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java [hunk 9]
```diff
@@ -55,133 +55,148 @@
...
+        ConnectionOfTravelAssessment event = new ConnectionOfTravelAssessment();
+        event.setIntersectionID(intersectionID);
+
+        doReturn(event).when(mongoTemplate).findOne(any(Query.class), eq(ConnectionOfTravelAssessment.class),
+                anyString());
+
+        Page<ConnectionOfTravelAssessment> page = repository.findLatest(intersectionID, startTime, endTime);
+
+        assertThat(page.getContent()).hasSize(1);
+        assertThat(page.getContent().get(0).getIntersectionID()).isEqualTo(intersectionID);
```
Comment (mcook42) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java]: suggestion(non-blocking): use the built-in method to get the first item in a collection
```suggestion
        assertThat(page.getContent().getFirst().getIntersectionID()).isEqualTo(intersectionID);
```

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/config/DefaultConfigRepositoryImplTest.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        List<DefaultConfig> result = repository.find(query);
+
+        assertThat(result).hasSize(1);
+        assertThat(result.get(0)).isEqualTo(config);
+        verify(mongoTemplate).find(eq(query), eq(DefaultConfig.class), eq("CmDefaultConfig"));
+    }
+
+    @Test
+    void testSaveIntegerType() {
+        DefaultConfig<String> config = mock(DefaultConfig.class);
```
Comment (mcook42) [inline services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/config/DefaultConfigRepositoryImplTest.java]: question(non-blocking): what's the benefit of mocking the DefaultConfig instead of constructing a real one like 
`DefaultConfig<String> config = new DefaultConfig<>("intKey", "category", "42", "java.lang.Integer", UnitsEnum.CENTIMETERS, "description");` (just as an example)?

**Review by mcook42 — APPROVED** [review]

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java (hunk 9 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/assessments/ConnectionOfTravelAssessmentRepositoryImplTest.java]: there are 165 cases of this, I will replace a few that look like this but leave the rest the same!

File: services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/config/DefaultConfigRepositoryImplTest.java (hunk 10 above)
Comment (jacob6838) [reply services/intersection-api/api/src/test/java/us/dot/its/jpo/ode/api/accessorTests/config/DefaultConfigRepositoryImplTest.java]: I just updated these to use actual examples of different message types. I was originally confused by the save method casting Strings, so didn't go with a full example. I do think the save method could be improved in the future.

**Review by drewjj — APPROVED** [review]
This builds and the tests all pass
